### PR TITLE
Add a Makefile option to disable libunwind

### PR DESCRIPTION
**Pull request by Keno**
There's been a request for a version of julia that does not depend on
libunwind, for easier bring up on new architectures. This adds a makefile
flag that can be put in Make.user (`DISABLE_LIBUNWIND`) and if set, will
disable any dependency on libunwind. Backtraces and profiling won't work
of course.

### CODE REVIEW

File: src/julia_internal.h [hunk 1]
```diff
@@ -597,7 +597,7 @@ static inline void jl_set_gc_and_wait(void)
 }
 #endif
 
-void jl_dump_native(const char *bc_fname, const char *obj_fname, const char *sysimg_data, size_t sysimg_len);
+void jl_dump_native(const char *bc_fname, const char *bc_unopt_fname, const char *obj_fname, const char *sysimg_data, size_t sysimg_len);
```
Comment (JeffBezanson) [inline src/julia_internal.h]: ?

Comment (Keno) [inline src/julia_internal.h]: Some unrelated changes crept in during rebase. Will take them out in the morning.

File: Make.inc [hunk 2]
```diff
@@ -26,6 +26,7 @@ OPENBLAS_USE_THREAD:=1
 # Use libraries available on the system instead of building them
 USE_SYSTEM_LLVM:=0
 USE_SYSTEM_LIBUNWIND:=0
+DISABLE_LIBUNWIND:=0
```
Comment (StefanKarpinski) [inline Make.inc]: We generally try to avoid double negatives in options, although in this case everything else defaults to false.

**Comment by ararslan** [comment]
@jlbuild !flags=DISABLE_LIBUNWIND=1 !nuke

**Comment by jlbuild** [comment]
Got it, building https://github.com/JuliaLang/julia/commit/1ae763626ca7aa4237c2f8d840af2bed5534f293

I will edit this comment as the build jobs complete.

**Comment by claudefalbriard** [comment]
When running the make under s390x, even having libunwind disabled, I receive the following error
at the build: 
 
make USE_SYSTEM_PCRE=1 PCRE= USE_SYSTEM_*=1 USE_SYSTEM_LIBM=1 USE_SYSTEM_BLAS=1 NO_LAPACK=1 USE_SYSTEM_LLVM=0 DISABLE_LIBUNWIND=1

```
checking for build architecture... s390x
checking for host architecture... s390x
checking for target architecture... s390x
checking for target operating system... linux-gnu
checking for ELF helper width... configure: error: Unknown ELF target: s390x
/data/shared/julia/julia-master/deps/unwind.mk:36: recipe for target 'scratch/libunwind-1.1-julia2/build-configured' failed
make[1]: *** [scratch/libunwind-1.1-julia2/build-configured] Error 1
Makefile:85: recipe for target 'julia-deps' failed
make: *** [julia-deps] Error 2
```

**Comment by Keno** [comment]
Pushed an update to disable it trying to download libunwind if the make option is set.

**Comment by claudefalbriard** [comment]
Just downloaded latest master distribution, but still getting the following error at the make step, using makefile flag DISABLE_LIBUNWIND=1

`make DISABLE_LIBUNWIND=1 USE_SYSTEM_PCRE=1 PCRE= USE_SYSTEM_*=1 USE_SYSTEM_LIBM=1 USE_SYSTEM_BLAS=1 NO_LAPACK=1 USE_SYSTEM_LLVM=0`

Error: 
`"checking for ELF helper width... configure: error: Unknown ELF target: s390x"`

Question: If the LIBUNWIND package is in a disabled state, why does it check the architecture?

**Comment by tkelman** [comment]
> latest master distribution

Are you building on this branch `kf/unwinddis` ? This pull request hasn't been merged yet.

**Comment by claudefalbriard** [comment]
@tkelman thanks for the hint. I've downloaded the branch and executed the make again. The process went much further, but stopped at a module called flisp. Looks like a dependency issue due to missing  libunwind files. Any advice how to fix this error? Here the error message: 
```
    LINK src/flisp/flisp
gcc: error: /data/shared/julia/julia-kf-unwinddis/usr/lib/libunwind-generic.a: No such file or directory
gcc: error: /data/shared/julia/julia-kf-unwinddis/usr/lib/libunwind.a: No such file or directory
Makefile:83: recipe for target '/data/shared/julia/julia-kf-unwinddis/src/flisp/flisp' failed
make[2]: *** [/data/shared/julia/julia-kf-unwinddis/src/flisp/flisp] Error 1
Makefile:224: recipe for target 'flisp/libflisp.a' failed
make[1]: *** [flisp/libflisp.a] Error 2
Makefile:55: recipe for target 'julia_flisp.boot.inc.phony' failed
make: *** [julia_flisp.boot.inc.phony] Error 2
```

**Comment by Keno** [comment]
I've updated this branch again to hopefully take care of that one as well.

**Comment by claudefalbriard** [comment]
@Keno , thanks for fixing this dependency. The latest downland of the branch and build process did throw the following error. Any advice how to correct this "core dumped" error at inference.ji?  

```
Warning: git information unavailable; versioning information limited
/bin/sh: 1: llvm-config: not found
    PERL base/build_h.jl.phony
WARNING: Using boilerplate git version info
Warning: git information unavailable; versioning information limited
    JULIA usr/lib/julia/inference.ji
Segmentation fault (core dumped)
Makefile:215: recipe for target '/data/shared/julia/julia-kf-unwinddis/usr/lib/julia/inference.ji' failed
make[1]: *** [/data/shared/julia/julia-kf-unwinddis/usr/lib/julia/inference.ji] Error 139
Makefile:100: recipe for target 'julia-inference' failed
make: *** [julia-inference] Error 2
```

**Comment by ViralBShah** [comment]
That error seems to now be beyond the scope of this issue.

**Comment by claudefalbriard** [comment]
@ViralBShah , I agree, certainty  latest build error (core dumped) is not related any-more to this branch, so I decided to report it under the issue "Julia Build from Source / s390x Machine Architecture not Supported" [ issue  [#21474](https://github.com/JuliaLang/julia/issues/21474) ].  Hints how to debug the error are welcome. Thanks to all colleagues for fixing the LIBUNWIND dependency.
